EPPs/common.py

```diff
@@ -385,6 +387,76 @@ def _run(self):
+    output_plate_layout_columns = None
+    input_plate_layout_columns = None
+    input_plate_layout_rows = None
+
+    def __init__(self, argv=None):
+        super().__init__(argv)
+        assert self.output_plate_layout_rows is not None, 'output_plate_layout_rows needs to be set by the child class'
+        assert self.output_plate_layout_columns is not None, 'output_plate_layout_columns needs to be set by the child class'
+        assert self.input_plate_layout_rows is not None, 'input_plate_layout_rows needs to be set by the child class'
+        assert self.input_plate_layout_columns is not None, 'input_plate_layout_columns needs to be set by the child class'
+
+    def generate_input_container_nested_dict(self):
+        # loop through the inputs, assemble a nested dicitonary {containers:{input.location:output} this can then be
+        # queried in the order container-row-column so the order of the inputs in the Hamilton input file is
+        # as efficient as possible.
+        nested_dict = {}
+        for art in self.artifacts:
+
+            # obtain outputs for the input that are analytes, assume step is not configured to allow replicates
+            # so will always work with output[0]
+            output = self.process.outputs_per_input(art.id, Analyte=True)[0]
+
+            # add the input_location_output_dict to the input_container_nested dict
+            if art.container not in nested_dict.keys():
+                nested_dict[art.container] = {}
+            # start assembling one of the variables needed to use the set_placement function
+            # note that .location returns a tuple with the container entity and the well location as the string in position [1]
+            nested_dict[art.container][art.location[1]] = output
+        return nested_dict
+
+    def generate_output_placement(self, output_container):
+
+        output_plate_layout = [(row + ":" + column) for column, row in
+                               itertools.product(self.output_plate_layout_columns, self.output_plate_layout_rows)]
+
+        placement = []
+        # obtain the dictionary containing the source information
+        input_container_nested_dict = self.generate_input_container_nested_dict()
+
+        # create a counter so only use each available well in the output plate once (24 wells available in the 96 well plate
+        well_counter = 0
+
+        # loop through the input containers and place the samples in row-column order - this makes pipetting as efficient
+        # as possible, particularly if only 1 input plate so 1:1 pipetting
+        for container in sorted(input_container_nested_dict, key=lambda x: x.name):
+            for column in self.input_plate_layout_columns:
+                for row in self.input_plate_layout_rows:
+                    # populate list of tuples for set_placements if well exists in input plate
+                    if row + ":" + column in input_container_nested_dict[container].keys():
```

Comment:
no need for the function call `.keys()` here

Reply:
updated